README
This directory contains the programs that convert the original web
sources to C.

The convert shell script is the thing that actually gets invoked from
the Makefile.  It runs web2c, fixwrites, and splitup in turn, plus
various sed nonsense.  web2c parses Pascal and outputs C; fixwrites
shouldn't be necessary, it converts multiple-argument writes into C;
splitup splits the output file into pieces for the sake of broken
compilers.

To define a new symbol (e.g., a function) to be used in the change
files, it's necessary to add it to texmf.defines (if it's only used in
TeX/MF/MP) or common.defines (otherwise) here, as well as making the
actual definition.  New functions in the library should be declared
somewhere, probably cpascal.h.  Of course no underscores are allowed,
since tangle removes them.




ChangeLog
2018-11-11  Andreas Scherer  <https://ascherer.github.io>

	* fixwrites.c (main): check return value from fgets.
	(tex-k 11 Nov 2018 11:53:24)

2018-06-08  Karl Berry  <karl@freefriends.org>

	* common.defines (aopeninwithdirname): add.

2016-04-06  Karl Berry  <karl@tug.org>

	* convert: remove $pascalfile if it is empty,
	along with moving the non-empty .c output, etc.

2015-06-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Integrate MFLua and MFLuaJIT.
	* coerce.h, convert, splitup.c: Adapted.

2015-04-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Test if isascii is declared.

2015-01-19  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines: Remove intcast.

2015-01-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines: Remove casttobytepointer and pointertobyte.

2014-11-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* configure.ac: Add AC_CHECK_SIZEOF([long]),
	AC_CHECK_SIZEOF([void *]) and AC_TYPE_UINTPTR_T.

2014-06-16  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Drop the obsolete ACLOCAL_AMFLAGS.

2014-01-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* fixwrites.c: Recognize dumpname as string.

2014-01-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Introduce a genuine 64-bit integer type, in case of 32-bit off_t.
	* common.defines (int64cast, integer64): Add these.

2013-12-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines (makesuffix): Add this.

2013-07-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (libweb2c.a): Avoid more 'parallel make' problems.

2013-07-15  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am (w2c/config.h): Avoid 'parallel make' problems.

2013-06-28  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* cvtbib.sed: Generate bibtex.c to first #include "cpascal.h".
	* makecpool.c: Generate tex-pool.c to first #include "texd.h".
	* web2c-lexer.l: First #include <w2c/config.h> via "web2c.h".

2013-06-27  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* main.c: Drop variable 'error' (set but not used).

2012-12-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Enable silent rules.

2012-11-20  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Avoid use of deprecated INCLUDES.

2012-08-22  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* fixwrites.c, main.c, makecpool.c, splitup.c:
	Use binary mode for output files.
	* fixwrites.c: Recognize (u)ppltotf as pltotf variant.
	From Akira Kakuto <kakuto@fuk.kindai.ac.jp> (W32TeX).

2012-07-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* web2c-parser.[ch]: Regenerate with bison 2.5.1.

2012-06-20  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* web2c-lexer.l: Added the never-interactive option to prevent
	flex from creating a problematic prototype for isatty().

2012-04-08  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* texmf.defines (callmakempx, getrandomseed): Removed, MP only.

2012-03-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Avoid 'differ in signedness' warnings.
	* common.defines: Define ustringcast().

2012-02-06  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* texmf.defines: Moved writedvi from here ...
	* common.defines: ... to here for use in GFtoDVI.

	* common.defines: Drop writechunk, GFtoDVI now uses writedvi.

2012-01-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* fixwrites.c: Replace 'fprintf(filename, "%c\n", char);' by
	'{ putc(char, filename); putc ('\n', filename); }'.  Patch from
	Akira Kakuto for (u)pDVItype & Co. where putc() => putc2().

2011-10-13  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines: Remove ALEPHHELP.

2011-10-04  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Add distclean-local to remove kpathsea/ and w2c/.

2011-08-17  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines: Remove ETEXHELP, LUATEXHELP, PDFTEXHELP.
	Add UPBIBTEXHELP, UPDVITYPEHELP, UPPLTOTFHELP, and UPTFTOPLHELP.
	* convert: Add upBibTeX, upDVItype, upPLtoTF, and upTFtoPL.

2011-08-12  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* convert, splitup.c: Add upTeX and e-upTeX.

2011-04-18  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Avoid 'differ in signedness' warnings.
	* common.defines: Define ucharcast().

2011-01-11  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* convert, splitup.c: Add e-pTeX.
	* splitup.c: #define onlyTeX in texd.h.

2010-06-10  Peter Breitenlohner  <peb@mppmu.mpg.de>

	Reorganize web2c headers.
	* native/kpathsea/c-auto.in (deleted): Moved ...
	* c-auto.in (new): ... to here.
	* configure.ac: Create kpathsea/c-auto.h instead of
	native/kpathsea/c-auto.h.
	* Makefile.am: Create w2c/config.h instead of native/config.h.
	Adjust INCLUDES, BUILT_SOURCES, DISTCLEANFILES, rules, and
	dependencies.
	* fixwrites.c, splitup.c, web2c.h: #include <w2c/config.h>
	instead of "config.h".

2010-04-26  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* Makefile.am: Make sure native/config.h exists before
	compiling web2c, fixwrites, or splitup.

2010-04-23  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* fixwrites.c (main): Recognize 'conststringcast...' as string.

	* Makefile.am (INCLUDES): Remove '-I$(srcdir)/..'.
	(native/config.h): Adapted.

2010-04-20  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* coerce.h: Remove luaTeX specific code.
	* convert: Reorganize (common code for TeX and MF).

2010-04-18  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* convert, splitup.c: Remove luatex.
	* common.defines: Remove LUATANGLEHELP.
	Add PBIBTEXHELP, PDVITYPEHELP, PPLTOTFHELP, and PTFTOPLHELP.

2010-04-11  Karl Berry  <karl@tug.org>

	* convert: handle pbibtex, ptex, ptftopl, ppltotf, pdvitype.
	* splitup.c: handle ptex.

2010-03-23  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines (BIBXRETALLOCSTRING): New procedure.

2010-03-09  Peter Breitenlohner  <peb@mppmu.mpg.de>

	* common.defines (BIBXRETALLOCNOSET): New procedure.

fixwrites.c
/* fixwrites -- convert Pascal write/writeln's into fprintf's or putc's.
   Originally by Tim Morgan, October 10, 1987.  */

#include <w2c/config.h>
#include <kpathsea/c-pathmx.h>

char buf[BUFSIZ], filename[PATH_MAX], args[100];
char *file, *argp, *as, *cmd;

int tex = false;

/* Replace the last (should be only) newline in S with a null.  */

static void
remove_newline (string s)
{
  char *temp = strrchr (s, '\n');
  if (temp == NULL)
    {
      fprintf (stderr, "Lost newline somehow.\n");
      /* This is so the convert script can delete the output file on error.  */
      puts ("@error@");
      exit (1);
    }

  *temp = 0;
}


static char *
insert_long (string cp)
{
  char tbuf[BUFSIZ];
  register int i;

  for (i = 0; &buf[i] < cp; ++i)
    tbuf[i] = buf[i];
  strcpy (&tbuf[i], "(long)");
  strcpy (&tbuf[i + 6], cp);
  strcpy (buf, tbuf);
  return cp + 6;
}


static void
join (string cp)
{
  char temp[BUFSIZ], *tp;

  if (!fgets (temp, BUFSIZ, stdin))
    return;
  remove_newline (temp);

  *cp++ = ' ';
  for (tp = temp; *tp == ' '; ++tp)
    ;

  strcpy (cp, tp);
}


static void
do_blanks (int indent)
{
  register int i;

  for (i = 0; i < indent / 8; i++)
    putchar ('\t');
  indent %= 8;
  for (i = 0; i < indent; i++)
    putchar (' ');
}


/* Return true if we have a whole write/writeln statement.  We determine
   this by matching parens, ignoring those within strings.  */
static int
whole (string cp)
{
  register int depth = 0;

  while (cp && *cp)
    {
      switch (*cp)
	{
	case '(':
	  ++depth;
	  break;
	case ')':
	  --depth;
	  break;
	case '"':
	  for (++cp; cp && *cp && *cp != '"'; ++cp)
	    if (*cp == '\\')
	      ++cp;
	  break;
	case '\'':
	  ++cp;
	  if (*cp == '\\') ++cp;
	  ++cp;
	  break;
	}
      ++cp;
    }
  return depth <= 0;
}


/* Skips to the next , or ), skipping over balanced paren pairs.  */

static char *
skip_balanced (string cp)
{
  register int depth = 0;

  while (depth > 0 || (*cp != ',' && *cp != ')'))
    {
      switch (*cp)
	{
	case '(':
	  ++depth;
	  break;
	case ')':
	  --depth;
	  break;
	}
      ++cp;
    }
  return cp;
}


/* Return true if c appears, except inside a quoted string */

static int
bare (string cp,  char c)
{
  for (; *cp && *cp != c; ++cp)
    {
      if (*cp == '"')
	{
	  ++cp;			/* skip over initial quotation mark */
	  while (*cp && *cp != '"')
	    {			/* skip to closing double quote */
	      if (*cp == '\\')
		++cp;
	      ++cp;
	    }
	}
      else if (*cp == '\'')
	{
	  ++cp;			/* skip to contained char */
	  if (*cp == '\'')
	    ++cp;		/* if backslashed, it's double */
	  ++cp;			/* skip to closing single-quote mark */
	}
    }
  return *cp;
}


/* xchr[...] is supposed to be replaced by Xchr(...)  when characters
   take more than a single octet each, as is the case in Aleph.  Now
   there are several occurrences of xchr[...[...]...], which are
   translated into Xchr(...[...)...], and the compiler dies on syntax
   errors.  Ensures that it is the matching bracket that is replaced,
   not the first one.  */

static char *
advance_cp (char *cp, int lefts)
{
  char *cp1;
  char *cp2;

  cp1 = strchr (cp + 1, ']');
  cp2 = strchr (cp + 1, '[');
  if (cp2 && cp2 < cp1)
    return advance_cp (cp2, lefts + 1);
  if (lefts == 1)
    return cp1;
  return advance_cp (cp1, lefts - 1);
}

#ifdef WIN32
#include <io.h>
#include <fcntl.h>
#endif

int
main (int argc,  string *argv)
{
  register char *cp;
  int blanks_done, indent, i;
  const char *program_name = "";

#ifdef WIN32
  setmode(fileno(stdout), _O_BINARY);
#endif
  for (i = 1; i < argc; i++)
    {
      if (STREQ(argv[i],"-t"))
	tex = true;
      else
	program_name = argv[i];
    }

  while (fgets (buf, BUFSIZ, stdin))
    {
      remove_newline (buf);
      blanks_done = false;

      for (cp = buf; *cp; ++cp) ;

      while (cp != buf && *--cp == ' ') ;

      while (*cp == '.')
	{
	  join (cp + 1);
	  while (*cp)
	    ++cp;
	  while (*--cp == ' ') ;
	}

      for (cp = buf, indent = 0; *cp == ' ' || *cp == '\t'; ++cp)
	{
	  if (*cp == ' ')
	    indent++;
	  else
	    indent += 8;
	}

      if (!*cp)
	{			/* All blanks, possibly with "{" */
	  puts (buf);
	  continue;
	}
      if (*cp == '{')

        {
	  do_blanks (indent);
	  putchar ('{');
	  ++cp;
	  while (*cp == ' ' || *cp == '\t')
	    ++cp;
	  blanks_done = true;
	  if (!*cp)
	    {
	      putchar ('\n');
	      continue;
	    }
	}

      if (!blanks_done)
	do_blanks (indent);

      if (strncmp (cp, "read ( input", 12) == 0)
	{
	  char variable_name[20];
	  if (sscanf (cp, "read ( input , %s )", variable_name) != 1)
            {
  	      fprintf (stderr, "sscanf failed\n");
              exit (1);
            }
	  printf ("%s = getint();\n", variable_name);
	  continue;
	}

      if (strncmp (cp, "lab", 3) == 0 && strchr (cp, ':'))
	{
	  do
	    {
	      putchar (*cp);
	    }
          while (*cp++ != ':');

          while (*cp == ' ')
	    ++cp;
	  putchar (' ');
	}

      if (strncmp (cp, "else write", 10) == 0)
	{
	  puts ("else");
	  do_blanks (indent);
	  cp += 5;
	  while (*cp == ' ')
	    ++cp;
	}

      if (bare (cp, '{'))
	{
	  while (*cp != '{')
	    {
	      putchar (*cp);
	      ++cp;
	    }
	  ++cp;
	  puts ("{");
	  indent += 4;
	  do_blanks (indent);
	  while (*cp == ' ')
	    ++cp;
	}

      if (strncmp (cp, "write (", 7) && strncmp (cp, "writeln (", 9))
	{
	  /* if not a write/writeln, just copy it to stdout and continue */
	  puts (cp);
	  continue;
	}
      cmd = cp;
      while (!whole (buf))	/* make sure we have whole stmt */
	{
	  if (!fgets (&buf[strlen (buf)], BUFSIZ - strlen (buf), stdin))
            break;
	  remove_newline (buf);
	}

      while (*cp != '(')
	++cp;
      ++cp;
      while (*(cp + 1) == ' ')
	++cp;

      /* Some writes start with a variable, instead of a file. */
      if (*(cp + 1) == '"' || *(cp + 1) == '\''
          || strncmp (cp + 1, "buffer", 6) == 0
          || strncmp (cp + 1, "xchr", 4) == 0
          || strncmp (cp + 1, "k ,", 3) == 0
          || strncmp (cp + 1, "s ,", 3) == 0
          || strncmp (cp + 1, "dig", 3) == 0
          || strncmp (cp + 1, "HEX", 3) == 0
          || strncmp (cp + 1, "versionstring", 13) == 0
          || strncmp (cp + 1, "kpathseaversionstring", 21) == 0
         )
	strcpy (filename, "stdout");
      else
	{
	  file = filename;
	  while (*cp != ',' && *cp != ')')
	    *file++ = *cp++;
	  *file = '\0';
	}
      if (*cp == ')')
	{
	  printf ("putc ('\\n', %s);\n", filename);
	  continue;
	}
      argp = ++cp;
      as = args;
      while (*cp == ' ')
	++cp;
      while (*cp != ')')
	{
	  if (*cp == '\'' || strncmp (cp, "xchr", 4) == 0
              || (strncmp (cp ,"HEX", 3) == 0
                  && (STREQ (program_name, "ofm2opl")
                      || STREQ (program_name, "opl2ofm")
                      || STREQ (program_name, "ovp2ovf")
                      || STREQ (program_name, "ovf2ovp")))
	      || strncmp (cp, "ASCII04", 7) == 0
	      || strncmp (cp, "ASCII1", 6) == 0
	      || strncmp (cp, "ASCIIall", 8) == 0              
	      || strncmp (cp, "months", 6) == 0
	      || strncmp (cp, "nameoffile", 10) == 0
	      || (strncmp (cp, "buffer", 6) == 0
                  && (STREQ (program_name, "vptovf")
                      || STREQ (program_name, "pltotf")
                      || STREQ (program_name, "ppltotf")
                      || STREQ (program_name, "uppltotf")
                      || STREQ (program_name, "ovp2ovf")
                      || STREQ (program_name, "opl2ofm")))
              || (((strncmp (cp, "buf", 3) == 0
		    || strncmp (cp, "xdig", 4) == 0
		    || strncmp (cp, "xext", 4) == 0
		    || strncmp (cp, "xhyf", 4) == 0)
                  && STREQ (program_name, "patgen")))
             )
	    {
	      *as++ = '%';
	      *as++ = 'c';
	      if (tex && strncmp (cp, "xchr", 4) == 0)
		{
		  *cp = 'X';
		  cp = strchr (cp, '[');
		  *cp = '(';
		  cp = advance_cp(cp,1);
		  *cp++ = ')';
		}
	      else if (*cp == '\'')
		cp += 2;
	    }
          
	  else if (*cp == '"')
	    {
	      *as++ = '%';
	      *as++ = 's';
	      while (*++cp != '"')	/* skip to end of string */
		if (*cp == '\\')
		  ++cp;		/* allow \" in string */
	    }

          /* More kludges -- recognize some things as strings and use %s:
             - versionstring
             - poolname
             - formatengine
             - dumpname */
          else if (strncmp (cp, "versionstring", 13) == 0
                   || strncmp (cp, "poolname", 8) == 0
                   || strncmp (cp, "formatengine", 12) == 0
                   || strncmp (cp, "dumpname", 8) == 0)
            {
              *as++ = '%';
              *as++ = 's';
            }

	  /* And yet another kludge, to handle stringcast (<whatever>) */
          else if (strncmp (cp, "stringcast", 10) == 0
                   || strncmp (cp, "conststringcast", 15) == 0)
	    {
	      *as++ = '%';
	      *as++ = 's';
	      cp = skip_balanced (cp);  /* Skip cast expression */
	    }

          else
	    {
	      *as++ = '%';
	      *as++ = 'l';
	      *as++ = 'd';
	      cp = insert_long (cp);
	      cp = skip_balanced (cp);	/* It's a numeric expression */
	    }
	  while (*cp != ',' && *cp != ')')
	    ++cp;
	  while (*cp == ',' || *cp == ' ')
	    ++cp;
	}

      if (strncmp (cmd, "writeln", 7) == 0)
	{
	  *as++ = '\\';
	  *as++ = 'n';
	}

      *as = '\0';
      if (strcmp (args, "%c") == 0)
	{
	  for (as = argp; *as; ++as) ;
	  while (*--as != ')') ;
	  *as = '\0';
	  printf ("putc (%s, %s);\n", argp, filename);
	}
      else if (strcmp (args, "%c\\n") == 0)
	{
	  for (as = argp; *as; ++as) ;
	  while (*--as != ')') ;
	  *as = '\0';
	  printf ("{ putc (%s, %s);  ", argp, filename);
	  printf ("putc ( '\\n', %s); }\n", filename);
	}
      else if (strcmp (args, "%s") == 0)
        printf ("Fputs (%s, %s\n", filename, argp);
      else
        printf ("fprintf (%s, \"%s\", %s\n", filename, args, argp);
    }

  return EXIT_SUCCESS;
}
